Model/DutyCalculator.php

```diff
@@ -116,84 +116,333 @@ public function __construct(
-                        $quote = $this->checkoutSession->getQuote();
-                        $baseCurrency = $this->storeManager->getStore()->getBaseCurrency();
-                        $rate = $baseCurrency->getRate($baseCurrency->getCode());
-                        $baseDuty = $duty / $rate;
-                        $quote->setBaseReachDuty($baseDuty);
-                        $quote->setReachDuty($duty);
-                        $quote->setDhlQuoteId($response['quoteId']);
-                        $quote->setDhlBreakdown(json_encode($response['feeTotals']));
-                        $quote->save();
-                    } else {
-                        $quote->setBaseReachDuty(0);
-                        $quote->setReachDuty(0);
-                        $quote->setDhlQuoteId('');
-                        $quote->setDhlBreakdown('');
-                        $quote->save();
+
+            //Countries (as per DHL doc) where both country name and states are required but are not selected yet.
+            //At present those are Canada and Brazil and few others (not sure what are those)
+            //What to do about those other countries?
+            if (in_array($address->getCountryId(), $special_countries, true)
+                && !$address->getRegionCode()
+            ) {
+                $this->handleStateUnspecifiedCase($duty);
+                return $this->response;
+            }
+
+            //previous country selection and current country selections are the same
+            //this check is necessary as during one order placement this place is entered multiple times due to the way
+            // ui event and data binding logic are set in Magento and our extension derived form that
+            if (($this->checkoutSession->getPrevCountry() == $address->getCountryId() )
+                && ( $address->getCountryId() !='')
+            ) {
+                $this->_logger->debug('Previous and current country selections are the same');
+                $this->response->setIsOptional($this->getIsOptional($address->getCountryId()));
+                //if chosen country is one of the special countries where state is needed
+                //but state is either not specified
+                if (in_array($address->getCountryId(), $special_countries, true)) {
+                    if (($this->checkoutSession->getPrevRegion() == $address->getRegionCode())
+                        || !$address->getRegionCode()  //this check is redundant here
+                    ) {
+                        $this->handleCaseWithCountryAndStateSpecified($address, $apply);
+                        return $this->response;
                     }
+                }
+                else {//country selection did not change and we are assuming (as we do not have extra info handy)
+                    //for these countries D&T does not changes with change in state (double check with the business)
```

Comment:
Miguel can double check with DHL but the duties and taxes do change based on region, so we'll need to send another request in the case that the country is the same but the region is different.  We can't take the previously calculated duty amount if the region has changed

Reply:
Miguel definitely can talk to DHL. 
If possible then let's talk about it tomorrow as well. Thanks!

Reply:
This piece is still outstanding - for what other countries, state/province name is relevant for accurate duty calculation.

Reply:
Okay so got the response from Miguel that for now state/province is relevant only for CA and BR as the countries of shipment. So current implementation done as part of MAG-70 is good in that area. 

Miguel  4:47 PM
Answer to the Canada / Brazil question - only those two at this time.
:heavy_check_mark:
1

Reply:
could you amend the comment in the code to clarify with the info we got from Miguel that for all other countries other than the ones in the special_countries list the state change doesn't matter? Just so that we don't have lingering q's that have been answered in the code

Reply:
Done as part of commit https://github.com/withreach/magento-plugin/pull/44/commits/b9d08fa382cc3a5934bca34cf7e1258d93a3b0f9
Thanks.